### PR TITLE
Moving iOS_system to dynamic libraries

### PR DESCRIPTION
**Pull request by holzschu**
As suggested by @IMcD23, all of ios_system is now based on dynamic libraries.
Visible impact on OpenTerm: null (for the time being). The dynamic libraries (lib*.dylib) must be compiled (one target each) and embedded, but do not need to be linked with the main program. Only iOS_system.framework needs to be linked. Adding/removing functionalities is done by adding/removing dynamic libraries, with no impact on the project file (easier to merge / add future functions). 

Pro: smaller memory footprint, faster launch.
Pro: easier to add new commands without having to change the Xcode project.
Con: swift functions must be mangled in order to be called by ios_system. Not impossible, but not exactly easy to read either.
Con: must add 7 dynamic libraries to the list of embedded libraries, +2 if you want Lua and Python, +6 if you want TeX.

Added bonus: two new commands, getoptString() and operatesOn(). Both take a command as argument. 
getoptString() returns the string that is sent to getopt(), that is the list of flags accepted by the command. For "rm", that would be "dfiPRrvW". 
operatesOn() returns what the command expects as arguments: "file", "directory" or "no". 

2nd bonus: a #define, SIDELOADING, which avoids defining commands that don't make sense on the app store: chgrp, chown, df... (I kept the existing list, but if we start writing scripts, we will need to make chmod available).

3rd bonus: Slight modifications to stdout in CommandExecutor.swift so commands know whether they're in a window or in a pipe, and react accordingly.

### CODE REVIEW

**Comment by louisdh** [comment]
Latest commit gives the following error on CI:
`ld: library not found for -lcrypto`

**Comment by holzschu** [comment]
I just saw that. Working on it.

**Comment by louisdh** [comment]
Also, can you take a look at the conflicts?
Please note that `pbcopy` and `pbpaste` have been added since you started working on this, so those need to be kept in the `viewDidLoad` function in `ViewController.swift`.

**Comment by louisdh** [comment]
Btw: I'm no fan of the mangled names used for the `replaceCommand` calls. Is there no way around that?

**Comment by holzschu** [comment]
I haven't found a way around the mangling, yet. Still looking for one.

**Comment by holzschu** [comment]
"terminal/Dependencies/libshell.dylib: No such file or directory"
I'm not sure I know how to fix that one.

**Comment by holzschu** [comment]
The mangling comes from the OO nature of Swift: several functions can have the same name, but different arguments. It's a way to differentiate them. One workaround would be to have C functions, with unique names, calling the Swift functions.

**Comment by IMcD23** [comment]
There’s no way that I know of to declare a C-style function in Swift, similar to extern “C”. There’s @convention(c) annotation, but I believe that only applies to blocks.

Perhaps expose these as @objc methods inside of a class, and refer to them by selector instead of providing a c method from ios_system.

Another option would be to pass a block/function ptr to ios_system.

**Comment by holzschu** [comment]
I think I have a way to mangle automatically. Because all functions have the same signature, the mangled name is basically identical: T08 + openTerm + length of function name + function name + "s5Int32VAD4argc_SpySpys4Int8VGSgGSg4argvtF". The last part meaning "(argc: Swift.Int32, argv: Swift.UnsafeMutablePointer<Swift.UnsafeMutablePointer<Swift.Int8>?>?) -> Swift.Int32"

And the 8 in T08 is the length of "openTerm". 
That can be encoded in a function. I will submit a revised version today.

**Comment by IMcD23** [comment]
@holzschu I think I may have a cleaner way to do `replaceCommand` that doesn't involve mangling.

```
// Define this somewhere shared.
public typealias SystemMain = @convention(c) (_ argc: Int32, _ argv: UnsafeMutablePointer<UnsafeMutablePointer<Int8>?>?) -> Int32

// Each global main function instead is a global constant, of type SystemMain
public let openUrl: SystemMain = { argc, argv in 
    // ...
}
```

Try that out and see if it works.

**Comment by holzschu** [comment]
@IMcD23 I tested what you suggested. The only result is that it changes the mangled name in the symbol table. 

The latest solution uses `mangleFunctionName("openUrl")`, which has good readability. The benefits of going through ios_system is that you get variable expansion and piping, which seemed to be important for openUrl.

**Comment by IMcD23** [comment]
Looks like conflicts in `CommandExecutor.swift` were caused by me with my 256 colors work. 

@holzschu You're now setting process-wide stdout while commands are running? Won't this cause issues if some other thread prints to stdout while a command is running?

**Comment by holzschu** [comment]
All commands write to thread local variable thread_stdout. However, thread_stdout must be initialized by the first command to stdout. This allows the command to know whether it is sending its output to the window (thread_stdout == stdout) or to another command, and adapt its response accordingly.  It is a replacement for the isatty() test.

**Comment by IMcD23** [comment]
@holzschu the reason I'm concerned about the stdout stuff is that I'd like to be able to support multiple concurrent commands in the future.

If you're setting app-wide stdout while a single command is running, then setting it back at the end, there will be issues if we implement a feature (which I'm working on right now) that allows multiple commands to run at once.

**Comment by holzschu** [comment]
I understand your concerns and your goals, but I fear before multiple commands can run simultaneously there are a *lot* of things to change in iOS_system, stdout being the last of them. 

Many Unix commands assume they're the only one running in their environment, because they're usually started with exec(), and they use global variables a lot. I did a good job of making them `static` and/or thread-local so they can run in sequence (in pipes, or with a commands calling system()). Making them run in parallel would require a *lot* more testing and substantial changes to the source of all commands.

Also, all commands access the same shell environment, by nature. How do you plan to give each tab its own environment? What will happen when two commands change the same environment variable? It can be done by rewriting `getenv()` and `setenv()` internally, but that's, again, a lot of work. 

Finally, ios_system relying on stdout is not a recent change. It's been like this since the beginning.

**Comment by IMcD23** [comment]
This makes a lot of sense, thanks for the explanation. I realized some of this after I implemented the tab stuff and was reading through the code, and ended up running all commands in the same serial queue, which guarantees they won’t execute concurrently. 

I don’t see a huge use case for multiple running at the same time, so I think this will be a good compromise.

The stdout/stderr issue is also fine 99% of the time, but I’ve still seen issues where UIKit constraint logging gets written to the terminal, and other logging that we don’t control. Need to figure out a way around that, but since it’s not a huge issue it’s not necessary to addres that in this change.

File: OpenTerm.xcodeproj/project.pbxproj [hunk 1]
```diff
@@ -592,12 +636,16 @@
...
 				ENABLE_BITCODE = NO;
+				FRAMEWORK_SEARCH_PATHS = (
+					"$(inherited)",
+					"$(PROJECT_DIR)/Dependencies",
+				);
 				INFOPLIST_FILE = "$(SRCROOT)/OpenTerm/Info.plist";
 				IPHONEOS_DEPLOYMENT_TARGET = 11.0;
 				LD_RUNPATH_SEARCH_PATHS = "$(inherited) @executable_path/Frameworks";
-                PRODUCT_BUNDLE_IDENTIFIER = com.silverfox.Terminal;
+				PRODUCT_BUNDLE_IDENTIFIER = "Nicolas-Holzschuch-OpenTerm";
```
Comment (louisdh) [inline OpenTerm.xcodeproj/project.pbxproj]: Please don't change the bundle identifier.

File: OpenTerm.xcodeproj/project.pbxproj [hunk 2]
```diff
@@ -613,12 +661,16 @@
...
```
Comment (louisdh) [inline OpenTerm.xcodeproj/project.pbxproj]: Same here.

File: OpenTerm.xcodeproj/project.pbxproj [hunk 3]
```diff
@@ -613,12 +661,16 @@
 				CODE_SIGN_ENTITLEMENTS = OpenTerm/OpenTerm.entitlements;
 				"CODE_SIGN_IDENTITY[sdk=iphoneos*]" = "iPhone Developer";
 				CODE_SIGN_STYLE = Automatic;
-                DEVELOPMENT_TEAM = 6G5LMQ72D8;
+				DEVELOPMENT_TEAM = VG8Z23C8YL;
```
Comment (louisdh) [inline OpenTerm.xcodeproj/project.pbxproj]: Please leave the development team as-is.

**Comment by holzschu** [comment]
Terribly sorry about that. I thought I had created a separate branch for my settings, so as not to pollute the master branch. Clearly I made a mistake.

**Comment by IMcD23** [comment]
Perhaps it’s best to squash the commits here, to not introduce your settings to history, and not pollute with 83 small commits.

**Comment by louisdh** [comment]
Thank you, is this ready to merge?

**Comment by holzschu** [comment]
I hope so.

**Pull request merged**